src/R/Components/Impl/Plots/Implementation/RPlotManager.cs

```diff
@@ -188,35 +283,143 @@ internal class RPlotManager : IRPlotManager {
+        private async Task ProcessPlotMessage(IRPlotDeviceViewModel viewModel, PlotMessage plot) {
+            if (plot.IsClearAll) {
+                await viewModel.PlotMessageClearAllAsync(plot.DeviceId, plot.DeviceNum);
+                History.RemoveAll(plot.DeviceId);
+            } else if (plot.IsPlot) {
+                try {
+                    var img = plot.ToBitmapImage();
+                    await viewModel.PlotMessageLoadPlotAsync(plot.DeviceId, plot.PlotId, img, plot.DeviceNum, plot.ActivePlotIndex, plot.PlotCount);
+                    History.AddOrUpdate(viewModel.DeviceName, viewModel.DeviceId, plot.PlotId, viewModel.SessionProcessId, img);
+                } catch (Exception e) when (!e.IsCriticalException()) {
+                    await viewModel.PlotMessageLoadErrorAsync(plot.DeviceId, plot.PlotId, plot.DeviceNum, plot.ActivePlotIndex, plot.PlotCount);
+                }
+            } else if (plot.IsError) {
+                await viewModel.PlotMessageLoadErrorAsync(plot.DeviceId, plot.PlotId, plot.DeviceNum, plot.ActivePlotIndex, plot.PlotCount);
+            }
+        }
 
-        public Task ExportToPdfAsync(string outputFilePath) =>
-            ExportAsync(outputFilePath, _interactiveWorkflow.RSession.ExportToPdfAsync(outputFilePath, PixelsToInches(_lastPixelWidth), PixelsToInches(_lastPixelHeight)));
+        private async Task RemoveFromHistoryAsync(Guid plotId) {
+            await InteractiveWorkflow.Shell.SwitchToMainThreadAsync();
+            History.Remove(plotId);
+        }
 
-        public void EndLocatorMode() {
-            EndLocatorMode(LocatorResult.CreateNotClicked());
+        private async Task ClearHistoryAsync() {
+            await InteractiveWorkflow.Shell.SwitchToMainThreadAsync();
+            History.Clear();
         }
 
-        public void EndLocatorMode(LocatorResult result) {
-            var tcs = _locatorTcs;
-            _locatorTcs = null;
-            tcs?.SetResult(result);
-            _interactiveWorkflow.Shell.DispatchOnUIThread(() => SetLocatorModeUI(false));
-            LocatorModeChanged?.Invoke(this, null);
+        private async void RSession_Mutated(object sender, EventArgs e) {
+            await RSessionMutatedAsync();
         }
 
-        private void SetLocatorModeUI(bool locatorMode) {
-            if (VisualComponent != null) {
-                VisualComponent.SetLocatorMode(locatorMode);
+        private async Task RSessionMutatedAsync() {
+            try {
+                var deviceId = await InteractiveWorkflow.RSession.GetActivePlotDeviceAsync();
+                var deviceChanged = ActiveDeviceId != deviceId;
+                ActiveDeviceId = deviceId;
+
+                // Update all the devices in parallel
+                var visualComponents = _visualComponents.Values.ToArray();
+                var tasks = visualComponents.Select(v => v?.ViewModel.RefreshDeviceNameAsync());
```

Comment:
Low-level perf perspective is the one I had in mind. There is some overhead in terms of the extra message header and WebSocket packet frame if you send more than one, of course, but it's on the order of tens of bytes - so negligible for large messages. And for small messages, it would only matter if you constantly send hundreds of them, which I don't think is the case here. 

The bigger problem is latency, when you serialize messages, i.e. send one, and then await before sending another. Send will return only when it got an ACK from the other side, so the channel ends up empty for the duration if nothing else is sending, wasting throughput - and this will be more noticeable on slow remote links. If you don't await on one send before sending another, you're making optimal use of the channel by saturating it. So for any higher-level APIs that we have that ultimately end up sending packets, avoiding unnecessary sequencing where there's no data dependency (i.e. avoiding `await` on individual calls as much as possible, and instead bundling them with `WhenAll` at points where you have to get the result), should be the optimal pattern for perf - and batching won't add much to it.